tck-flow/src/test/java/org/reactivestreams/tck/flow/SubmissionPublisherTest.java

```diff
@@ -0,0 +1,60 @@
+/************************************************************************
+ * Licensed under Public Domain (CC0)                                    *
+ *                                                                       *
+ * To the extent possible under law, the person who associated CC0 with  *
+ * this code has waived all copyright and related or neighboring         *
+ * rights to this code.                                                  *
+ *                                                                       *
+ * You should have received a copy of the CC0 legalcode along with this  *
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.*
+ ************************************************************************/
+
+package org.reactivestreams.tck.flow;
+
+import org.reactivestreams.tck.TestEnvironment;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.*;
+
+@Test
+public class SubmissionPublisherTest extends FlowPublisherVerification<Integer> {
+
+    public SubmissionPublisherTest() {
+        super(new TestEnvironment());
+    }
+
+    @Override
+    public Flow.Publisher<Integer> createFlowPublisher(final long elements) {
+        final SubmissionPublisher<Integer> sp = new SubmissionPublisher<Integer>();
+
+        Thread t = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (sp.getNumberOfSubscribers() == 0) {
+                    try {
+                        Thread.sleep(1);
+                    } catch (InterruptedException ex) {
+                        return;
+                    }
+                }
+
+                for (int i = 0; i < elements; i++) {
+                    sp.submit(i);
```

Comment:
Does this stop blocking if all subscribers cancel their subscriptions?

Reply:
I guess so. This test passes for me:

```java
        SubmissionPublisher<Integer> sp = new SubmissionPublisher<>(
                ForkJoinPool.commonPool(), 1);

        TestConsumer<Integer> tc1 = new TestConsumer<>(0L);
        TestConsumer<Integer> tc2 = new TestConsumer<>(0L);

        sp.subscribe(tc1);
        sp.subscribe(tc2);

        CountDownLatch cdl = new CountDownLatch(1);

        SchedulerServices.single().schedule(() -> {
            for (int i = 0; i < 512; i++) {
                sp.submit(i);
            }
            cdl.countDown();
        });

        Thread.sleep(1000);

        assertEquals(1, cdl.getCount());

        tc1.cancel();
        tc2.cancel();

        assertTrue(cdl.await(5, TimeUnit.SECONDS));
```